### ISSUE TITLE
Intl.PluralRules()

### ISSUE THREAD

**Issue by zbraniecki** [tc39/ecma402#34]
UPDATE: This proposal has advanced to stage 1, details here: https://github.com/caridy/intl-plural-rules-spec

cloned from: https://groups.google.com/forum/#!topic/javascript-globalization/3nFDf5al5hU

It would be very helpful for all l10n libraries to get CLDR plural forms into Intl API.

The proposed API could look like this:

```
var cardinal = new Intl.PluralFormat(‘en’, {style: ‘cardinal’});
console.log(cardinal.format(0)); // “other”
console.log(cardinal.format(1)); // “one”
console.log(cardinal.format(2)); // “other”

var ordinal = new Intl.PluralFormat(‘en’, {style: ‘ordinal’});
console.log(ordinal.format(11)); // “one”
console.log(ordinal.format(22)); // “two”
console.log(ordinal.format(33)); // “few”
console.log(ordinal.format(44)); // “other”
```

**Comment by caridy** [comment]
Makes sense, and it aligns well with some other proposals around this. Last time we talked about ICU messages, we ended up agreeing that better to do baby steps on this and get pluralization in place instead.

/cc @ericf, he has some ideas around this, and he was planning to champion the pluralization low level api.

**Comment by caridy** [comment]
Here is a backpointer to the discussion: https://groups.google.com/forum/#!topic/javascript-globalization/3nFDf5al5hU

**Comment by ericf** [comment]
@caridy Yeah @zbraniecki took this from the email thread. You guys both posted the same link to it :)

**Comment by caridy** [comment]
jaja

**Comment by zbraniecki** [comment]
The latest proposal that seems to address all comments so far:

```
var f = Intl.PluralForm(navigator.languages);

f.resolve(100); // 'other'
f.format(100, {
  one: 'comment',
  other: 'comments'
}); // 'comments'
```

I would say that the latter method should be lower priority for us. I like how simple it is, but I'm not sure if it would really be applicable for libraries that need it, while the former is definitely going to be directly applicable in multiple libraries I know of.

**Comment by caridy** [comment]
@zbraniecki let's keep the `format` with options out of the initial scope of this feature, we can discuss more later on.

**Comment by caridy** [comment]
Ok guys, finally got a chance to work on this, the initial proposal flush out:

https://github.com/caridy/intl-plural-form-spec

The generated spec [draft] is here:
https://rawgit.com/caridy/intl-plural-form-spec/master/index.html

I will work with @ericf next week for formalize the proposal and to see if we can formally presented (if we get a chance to add it to the agenda, which I doubt).

_P.S.: I hate to post in groups and here._

**Comment by srl295** [comment]
Yes, just resolve plurals first

> I hate to post in groups and here

Just post a link to here and say follow up in github?

**Comment by srl295** [comment]
The spec looks great, very straightforward.

**Comment by eemeli** [comment]
Thanks for the pointer @rxaviers. I added my comments to the mailing list [thread](https://groups.google.com/d/msg/javascript-globalization/3nFDf5al5hU/Lles9SToCQAJ).

Once the spec is ready-ish, I could probably pretty easily implement a polyfill for `Intl.PluralRules` from [make-plural.js](/eemeli/make-plural.js).

**Comment by ericf** [comment]
@eemeli awesome! I was hoping you'd say that :smile:

**Comment by caridy** [comment]
@eemeli let's have the discussion about your feedback here https://github.com/caridy/intl-plural-rules-spec/issues/1

**Comment by rxaviers** [comment]
@eemeli thx

**Comment by caridy** [comment]
_update: we are cooking the proposal for the next meeting in two weeks._

**Comment by eemeli** [comment]
Forgot to mention it here, but I did actually end up writing a [polyfill for Intl.PluralRules](/eemeli/IntlPluralRules). There are some comments about how it diverges from the current spec at caridy/intl-plural-rules-spec#3

**Comment by caridy** [comment]
we have reached stage 1 on this proposal :)

**Comment by rxaviers** [comment]
:tada:

**Comment by zbraniecki** [comment]
Heads up, this has been advanced to Stage 2 as of today. @ericf is championing it and what's left is to update the spec to deal with decimal portion of the Number. We have reviewers assigned.

**Comment by zbraniecki** [comment]
Slides from TC39 meeting: https://docs.google.com/presentation/d/1WflzD4YMOYYmJDAmjwmG0wi5GnLZd74EwvDRKSL9vA8/edit#slide=id.gd4949377b_1_17

**Comment by littledan** [comment]
A few feature requests came up in discussion among some of us working on ECMA-402. I'm not sure if they might already be reflected in the text, but just to record the discussion, here's what we've talked about:
- More options should be added to PluralRules, in parallel with the options in NumberFormat which affect how the number is rounded and how many decimal digits are given. These include `minFractionalDigits`, `maxFractionalDigits`, and similar for signficant digits. This solves the 1.0 
- format should be renamed to select, which is what CLDR calls it and which would reduce confusion since it doesn't produce the kind of string a user should see directly
- If possible, the CLDR range type should be exposed. This could be in the form of a `type: 'range'`, or maybe better because it's more basic, a separate `selectRange` method which takes two different types as arguments and returns the combined type based on per-language rules. Maybe we would want both, as a convenience method, or maybe stay minimal at first.
- The slides should add an example use case, something simple like
```js
var pr = new Intl.PluralRules('en', { type: 'cardinal' })
switch (pr.select(beers)) {
  case 'one': return 'one beer';
  default: return `${beers} beers`;
}
```

**Comment by eemeli** [comment]
Great!

However, the [current draft](https://rawgit.com/caridy/intl-plural-rules-spec/88a99f5/index.html) still has the issue of not dealing correctly with the categorisation of numerical strings with visible fractional digits that end in `0`. The unicode [LDML](http://unicode.org/reports/tr35/tr35-numbers.html#Operands) uses the operands `v` and `f` for these, and they're actually pretty common.

For a specific example, in English you have `1 widget`, but `1.0 widgets`.

To fix this, [ResolvePlural](https://rawgit.com/caridy/intl-plural-rules-spec/master/index.html#sec-ResolvePlural) should accept string input that it asserts to contain the string representation of a number, and the `select()` call should not cast such a string to a number before passing it on to ResolvePlural.

Or is this what you meant by the "decimal portion of the Number"?

**Comment by caridy** [comment]
that is correct @eemeli, that's what @littledan meant by "decimal portion of the Number", we have some ideas about how to solve that in the spec text, we will be working on that very soon.

**Comment by zbraniecki** [comment]
@littledan - what's the update on this proposal? Do you still to plan to get it to reviewers before the next ECMA meeting?

**Comment by littledan** [comment]
@zbraniecki I think that's more of a question for the champions. I'm happy to review it if it's ready; is it?

**Comment by zbraniecki** [comment]
Oh, sure. @ericf - any updates?

**Comment by ericf** [comment]
@zbraniecki No updates from me, I haven't been working on it.

**Comment by zbraniecki** [comment]
Ok!
@caridy - I created https://github.com/caridy/intl-plural-rules-spec/pull/4 which adds `GetOperands` based on http://unicode.org/repos/cldr/trunk/specs/ldml/tr35-numbers.html#Operands

In my testing that allowed for nice fractional digit resolution.

**Comment by eemeli** [comment]
I also added caridy/intl-plural-rules-spec#5, which drops the forced cast to Number that would make @zbraniecki's `GetOperands` to always return 0 for the `v` and `f` operands.

**Comment by zbraniecki** [comment]
This proposal has been submitted to reviewers and we hope to be able to advance to Stage 3 at TC39 meeting in Munich in May.

**Comment by zbraniecki** [comment]
We should update the Intl.js implementation to match the spec.

I already updated the `dayPeriod` and `literal` token names, but in the spec we changed function names and made the `formatToParts` non-bound.

**Comment by zbraniecki** [comment]
This proposal has been advanced to Stage 3. There's still editorial work going on in https://github.com/tc39/proposal-intl-plural-rules/pull/15 , but the API is ready and I started working on SpiderMonkey implementation (behind the flag). I think we may want to start working on Intl.js one and test262 tests soon.

**Comment by zbraniecki** [comment]
@eemeli - I started working on an implementation of the spec for Intl.js. One item that I could use help with is getting the data from CLDR into the lib. Did you solve it already and can you port it to Intl.js?

**Comment by eemeli** [comment]
In [make-plural.js](/eemeli/make-plural.js) I use Unicode's own [cldr-core](/unicode-cldr/cldr-core) JSON export of the official XML. Is that what you're looking for?

**Comment by zbraniecki** [comment]
Opened an issue in https://github.com/andyearnshaw/Intl.js/issues/243

**Comment by zbraniecki** [comment]
I landed `mozIntl.PluralRules` today in mozilla-central: https://hg.mozilla.org/mozilla-central/rev/80234a27490b
which is a full implementation of PluralRules spec proposal.

@jungshik, @littledan - is there a chance we could get it in V8 anytime soon? 
I know jungshik said that it should be easy - https://bugs.chromium.org/p/v8/issues/detail?id=5601 - but now it's a bit of a race against the time. If we can get Google's support behind it in time for January TC39 meeting, I'd like to suggest it for stage 4.

**Comment by littledan** [comment]
We have a bug open for it, but we don't have an implementation yet. I'd like to avoid thinking about these things as a race against time. It doesn't seem likely that this will get into ES2017, but the annual version numbers shouldn't matter for most purposes. Instead, what should matter is when it gets to Stage 4, meeting by meeting, so there's no real significant cutoff cliff. Chrome has shipped a number of features when they were at Stage 4 without waiting for the annual numbered version, e.g., async/await.

I've been making some changes to the Intl implementation recently to try to make our ECMA 402 implementation have less technical debt so it easier to add more features. We have this bug filed. Many people are taking Christmas break. Let's be OK with this taking a little bit of time.

**Comment by zbraniecki** [comment]
> Let's be OK with this taking a little bit of time.

As I mentioned in the other comment, my understanding was that ES releases are creating opportunity to increase adoption of ECMA402 features.

> Instead, what should matter is when it gets to Stage 4, meeting by meeting

Understood.

**Comment by ray007** [comment]
Any reason `ordinal`/`cardinal`/`spell-out` doesn't become a style for `Intl.NumberFormat`?

**Comment by littledan** [comment]
@ray007 Can you say more about your use case?

**Comment by ray007** [comment]
Our customers can configure "number controls" which allows to select 1 of several styles.
The formatted number is part of the displayed string.
The ICU number formatting API also seems to have styles for ordinal, cardinal, ...
I can work with it either way, as long as it gets implemented somewhere.

**Comment by rxaviers** [comment]
@ray007 are you saying `Intl.NumberFormat` should be used to output plural forms? For example, `new Intl.NumberFormat('en', {style: 'ordinal'}).format(0); // "other"`? Can you explain your suggestion?

**Comment by ray007** [comment]
Explain?
Not sure what kind of explanation/reasoning you would expect for an opinion?
It just seems easy to extend the existing `NumberFormat` and I had thought it would maybe take less discussion as making a new Object...

**Comment by rxaviers** [comment]
Sorry, but your suggestion is pretty vague to me and you're not helping to clarify it.

**Comment by ray007** [comment]
To use your example:
```
new Intl.NumberFormat('en', {style: 'ordinal'}).format(1);
```
should return `1st` or maybe a variation with "st" in superscript.

**Comment by rxaviers** [comment]
Thanks, I know understand what you want. Though, I'm afraid such feature isn't the right way to go. Please, see this reply from Mark Davis (Unicode president) about this topic: http://unicode.org/pipermail/cldr-users/2016-February/000462.html

**Comment by zbraniecki** [comment]
ordinal number formatting has nothing to do with PluralRules.

**Comment by zbraniecki** [comment]
This proposal has advanced to Stage 4 at today's TC39 meeting.

The committee raised the question about rationale for `type: ordinal`. One question was if it would be possible to support this via NumberFormat that would format `2nd`, `3rd` etc. without the need to adjust the whole sentence to the plural category. That seems to be the case in English.

Using Fluent syntax:

```
key = { PLURAL($num, type="ordinal") ->
    [one] {$num}st floor
    [two] {$num}nd floor
    [few] {$num}rd floor
    [other] {$num}th floor
}
```

vs.

```
key = { NUMBER($num, type="ordinal") } floor
```

But we identified a number of languages that require a whole sentence to be adjusted based on the ordinal plural category.

Examples:
* Filipino: http://www.unicode.org/cldr/charts/29/supplemental/language_plural_rules.html#fil
* Irish: http://www.unicode.org/cldr/charts/29/supplemental/language_plural_rules.html#ga
* Welsh: http://www.unicode.org/cldr/charts/29/supplemental/language_plural_rules.html#cy

For example in Filipino it seems to require:

```
key = { PLURAL($num, type="ordinal") ->
    [one] Lumiko sa unang kanan.
    [other] Lumiko sa ika-{$num} kanan.
}
```

while Irish requires:

```
key = { PLURAL($num, type="ordinal") ->
    [one] Glac an {$num}ú chasadh ar dheis
    [other] Glac an {$num}ú casadh ar dheis.
}
```

Thus we decided to stay with the current proposal and advance it to Stage 4!

**Comment by caridy** [comment]
@littledan are you going to send the PR to merge the new spec into 402? or do you need some help?

**Comment by mathiasbynens** [comment]
V8 v6.3.172 and Chrome 63 ship it as well. https://developers.google.com/web/updates/2017/10/intl-pluralrules

**Comment by jackhorton** [comment]
Joining the party a bit late here, but the PR to add PluralRules to ChakraCore has opened and should land soon (coming in Edge later this year): https://github.com/Microsoft/ChakraCore/pull/4940

**Comment by eemeli** [comment]
As a sidenote, I recently updated my polyfill to match the current spec: https://github.com/eemeli/intl-pluralrules
